### ISSUE TITLE
Errors after upgrading from 2.4.4 to 2.5 and higher

### ISSUE THREAD

**Issue by GinoPane** [neo4j-php/neo4j-php-client#146]
**Describe the bug**

Could you please help with neo4j-php-client issues? Once migrated from 2.4.4 to 2.5.0 the client gives timeout and thread limit errors for the same DB. So it is working fine on 2.4.4 and does not work from 2.5.0 and above

**To Reproduce**
1. Checkout https://github.com/AutoProtect-Group/neo4j-php-ogm ^2.1.6
2. Change neo4j-client requirement in composer.json to ^2.5 or higher
4. Run unit tests
5. See error

**Expected behavior**
There are no errors on the version 2.4.4. Tests are running fine

**Comment by transistive** [comment]
Thanks @GinoPane! I'll take a look at it now

**Comment by transistive** [comment]
Hi Gino,

It does indeed look very weird. Version 2.5 contained some optimisations, but it might bring everything out of whack in Neo4j 3.5. I'm trying to test it on Neo4j 4, but it does not recognize that it needs to use ${} bracket style. 

Can you guide me on how to turn on Neo4j 4 behaviour? Comparing how it deals with both versions will make it easier to debug

Thanks!

**Comment by GinoPane** [comment]
@transistive we will provide a branch with Neo4j 4.4 enabled

**Comment by KudrautsauSiarhei** [comment]
I forked the repo and created a branch, where updated neo4j to 4.4. You can use it:
https://github.com/KudrautsauSiarhei/neo4j-php-ogm/tree/my_branch

**Comment by transistive** [comment]
Thanks @KudrautsauSiarhei,

It looks like I don't properly close off the connections, so it'll take me some time to figure it out. I know how I am going to spend my weekend :partying_face: 

Thank you all for posting this issue and for your patience. This is good stuff

**Comment by GinoPane** [comment]
@transistive thanks! Looking forward to hearing from you.

**Comment by KudrautsauSiarhei** [comment]
@transistive Hi, have you found anything yet? Maybe you already found the error?

**Comment by transistive** [comment]
Yes @KudrautsauSiarhei,

There were actually a few inconsistencies which I was already able to solve.

As luck would have it I was already working on a rework of the connection manager which is where the mistake happens. You should see a PR pop up later during the day!

Thank you for your patience

**Comment by transistive** [comment]
So the PR is still in a draft phase. Is it OK if I use the rest of the week to finish it, as it encompasses quite a bit of changes? The bug you guys are experiencing is mainly because it leaves too many open connections, and the current connection pool implementation doesn't handle it correctly.

I have been thinking about solving the connection pooling problem once and for all by allowing users to optionally install an implementation of a PSR simple cache, which makes it possible to ensure an application never opens too many connections for Neo4j to handle, even when using separate processes or PHP sessions.

**Comment by GinoPane** [comment]
@transistive is there anything we can do to help you with this?

**Comment by transistive** [comment]
Thank you for the offer! - There is, actually.

It would be a great help if you can simplify the test that breaks, so it is reproducible with the client only. If you send a PR with a simplified test, we can be sure the problem is solved.

**Comment by GinoPane** [comment]
@transistive do you mean the test for neo4j-php-client project, right?

**Comment by transistive** [comment]
Yes :+1:,

I'll be working on some other projects today, but will circle back to this tomorrow and tonight

**Comment by KudrautsauSiarhei** [comment]
@transistive 
Can you tell me what should be in the NEO4J_CONNECTIONS environment variable for the local environment?

**Comment by transistive** [comment]
Yes, it's supposed to be a comma seperated list of uri's to connect to. (eg: bolt://localhost,http://neo4j:test@neo4j:7475)

**Comment by KudrautsauSiarhei** [comment]
When I run the tests locally I get several errors. I'm trying to understand why this is. Maybe because I incorrectly specified the values in NEO4J_CONNECTIONS?
The command I use to run tests: `docker-compose exec testkit-backend ./vendor/bin/phpunit tests`
![image](https://user-images.githubusercontent.com/91125404/183643289-4883a533-5513-4018-ad1d-87819c429717.png)

**Comment by transistive** [comment]
Does it work if you use client instead of testkit backend?

**Comment by transistive** [comment]
Testkit backend does not use the .env explicitly meaning you have to load it manually during bootstrapping (eg bootstrap_override_with_dotenv.php used by some workers of github actions)

**Comment by KudrautsauSiarhei** [comment]
I ran the client: `docker-compose run client bash`
And then tests: `./vendor/bin/phpunit tests`
But I still get some errors. I am running tests on tag `2.7.0`.
I took the value for `NEO4J_CONNECTIONS` from the example.
![image](https://user-images.githubusercontent.com/91125404/183646806-7de905fe-6773-4c66-aa6b-fa5a12e95c2d.png)

**Comment by transistive** [comment]
Very strange,

I can run the tests without any problems. I just double-checked. This is a great error though, are you using a different setup with Neo4j or did you just run `docker-compose up` ?

**Comment by KudrautsauSiarhei** [comment]
I just used docker-compose up
![image](https://user-images.githubusercontent.com/91125404/183648520-79f5577d-0b05-4203-a05a-21212ffbdf88.png)

**Comment by transistive** [comment]
Does it keep giving not a leader errors on subsequent reruns?

**Comment by KudrautsauSiarhei** [comment]
It still gives the same errors when restarting tests. Maybe it's because I'm running this inside wsl?
I have attached the full error log. Have you tried deleting your composer.lock and reinstalling the dependencies?
[error.txt](https://github.com/neo4j-php/neo4j-php-client/files/9290513/error.txt)

**Comment by transistive** [comment]
Got it. The NEO4J_CONNECTIONS are misconfigured

Bolt does not autoroute queries, so it tries to reach core1, which may or may not be a leader at that moment.

You need to switch bolt and neo4j schemes. The bolt scheme is only guaranteed to work on the standalone neo4j server

**Comment by transistive** [comment]
Can you try it with this one? `NEO4J_CONNECTIONS=bolt://neo4j:test@neo4j,neo4j://neo4j:test@core1,http://neo4j:test@neo4j`

**Comment by transistive** [comment]
Also make sure to reboot all docker stuff, just in case

**Comment by KudrautsauSiarhei** [comment]
It worked, thanks. Now I have only 38 skipped tests and 1 error test, the rest were completed successfully.

Tests: 410, Assertions: 302596, Errors: 1, Skipped: 38.
```
There was 1 error:

1) Laudis\Neo4j\Tests\Performance\PerformanceTest::testMultipleTransactions
fwrite(): Send of 3 bytes failed with errno=32 Broken pipe

/opt/project/vendor/stefanak-michal/bolt/src/connection/StreamSocket.php:85
/opt/project/vendor/stefanak-michal/bolt/src/protocol/AProtocol.php:48
/opt/project/vendor/stefanak-michal/bolt/src/protocol/V3.php:172
/opt/project/src/Bolt/BoltConnection.php:170
/opt/project/src/Bolt/BoltConnection.php:398
/opt/project/vendor/kriswallsmith/buzz/lib/Client/FileGetContents.php:31
/opt/project/src/Http/HttpConnectionPool.php:95
/opt/project/src/Http/HttpSession.php:110
/opt/project/src/Client.php:76
/opt/project/src/Client.php:69
/opt/project/src/Client.php:64
/opt/project/tests/Performance/PerformanceTest.php:132
/opt/project/tests/Performance/PerformanceTest.php:73

Caused by
fwrite(): Send of 3 bytes failed with errno=32 Broken pipe

/opt/project/vendor/stefanak-michal/bolt/src/connection/StreamSocket.php:85
/opt/project/vendor/stefanak-michal/bolt/src/protocol/AProtocol.php:48
/opt/project/vendor/stefanak-michal/bolt/src/protocol/V3.php:172
/opt/project/src/Bolt/BoltConnection.php:170
/opt/project/src/Bolt/BoltConnection.php:398
/opt/project/vendor/kriswallsmith/buzz/lib/Client/FileGetContents.php:31
/opt/project/src/Http/HttpConnectionPool.php:95
/opt/project/src/Http/HttpSession.php:110
/opt/project/src/Client.php:76
/opt/project/src/Client.php:69
/opt/project/src/Client.php:64
/opt/project/tests/Performance/PerformanceTest.php:132
/opt/project/tests/Performance/PerformanceTest.php:73
```

I also tried changing the value of this variable to something like this:
`NEO4J_CONNECTIONS=neo4j://neo4j:test@neo4j`
And there were only 29 tests skipped and there were no errors.
Tests: 246, Assertions: 101998, Skipped: 29.

**Comment by KudrautsauSiarhei** [comment]
@transistive I've added some tests for bugs that break our code.
https://github.com/neo4j-php/neo4j-php-client/pull/148

**Comment by transistive** [comment]
Super! I'll merge them with the routing table implementation

**Comment by GinoPane** [comment]
Hi @transistive! How is development going?
Can we maybe help somehow?

**Comment by GinoPane** [comment]
Hi @transistive! Any news about this one?

**Comment by transistive** [comment]
hello @GinoPane, I forgot to mention this but I fixed it over a month ago in version 2.7.0, Can you confirm it works?

**Comment by GinoPane** [comment]
@transistive yep, thanks! It worked
